File: latest/week3/workshop8_m4/W08_M04_demo.R
# ============================================================
# Workshop 8 — Module 4: Shiny Fundamentals
# Integrated R Script — UNYT Tirana
# ============================================================
# This script contains TWO Shiny apps and supporting code.
# Run each app section individually (not the whole script).
library(tidyverse); library(shiny); library(plotly); library(DT)
dir.create("output", showWarnings = FALSE)

# ── DATA LOADING (shared across apps) ────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")),
    duration_num = parse_number(duration))

# ══════════════════════════════════════════════════════════
# APP 1: MINIMAL SHINY — Netflix Explorer
# Demonstrates: selectInput, sliderInput, renderPlot, reactive
# ══════════════════════════════════════════════════════════

app1_ui <- fluidPage(
  titlePanel("Netflix Content Explorer"),
  sidebarLayout(
    sidebarPanel(width = 3,
      # Input 1: Content type dropdown
      selectInput("type", "Content Type:",
        choices = c("All", "Movie", "TV Show"),
        selected = "All"),

      # Input 2: Year range slider
      sliderInput("years", "Year Range:",
        min = 2015, max = 2021,
        value = c(2015, 2021), step = 1, sep = ""),

      # Input 3: Country dropdown
      selectInput("country", "Country:",
        choices = c("All", sort(unique(na.omit(nf$primary_country)))),
        selected = "All"),

      # Summary text
      textOutput("n_titles"),
      hr(),
      # Reset button
      actionButton("reset", "Reset All Filters", icon = icon("refresh"))
    ),
    mainPanel(width = 9,
      # Tab panels for different views
      tabsetPanel(
        tabPanel("Trend", plotOutput("trend_plot", height = "400px")),
        tabPanel("Interactive", plotlyOutput("plotly_chart", height = "400px")),
        tabPanel("Data Table", DTOutput("data_table"))
      )
    )
  )
)

app1_server <- function(input, output, session) {
  # ── REACTIVE: Filtered data ────────────────────────────
  # This recalculates automatically when any input changes
  filtered <- reactive({
    df <- nf |>
      filter(!is.na(year_added),
        year_added >= input$years[1],
        year_added <= input$years[2])

    if (input$type != "All") df <- df |> filter(type == input$type)
    if (input$country != "All") df <- df |> filter(primary_country == input$country)
    df
  })

  # ── OUTPUTS ────────────────────────────────────────────
  # Text: number of titles
  output$n_titles <- renderText({
    paste0("Showing ", format(nrow(filtered()), big.mark = ","), " titles")
  })

  # Static ggplot
  output$trend_plot <- renderPlot({
    filtered() |>
      count(year_added, type) |>
      ggplot(aes(x = year_added, y = n, color = type)) +
      geom_line(linewidth = 1.5) + geom_point(size = 3) +
      scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
      theme_minimal(base_size = 14) +
      theme(panel.grid = element_blank(),
        legend.position = "bottom") +
      labs(title = "Yearly Additions by Type",
           x = "Year", y = "Titles Added", color = NULL)
  })

  # Interactive plotly
  output$plotly_chart <- renderPlotly({
    p <- filtered() |>
      count(year_added, type) |>
      ggplot(aes(x = year_added, y = n, color = type,
        text = paste0("Year: ", year_added, "\nType: ", type, "\nCount: ", n))) +
      geom_line(linewidth = 1) + geom_point(size = 2) +
      scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
      theme_minimal() +
      labs(x = "Year", y = "Titles", color = NULL)
    ggplotly(p, tooltip = "text") |>
      layout(hovermode = "x unified", legend = list(orientation = "h"))
  })

  # Data table
  output$data_table <- renderDT({
    filtered() |>
      select(title, type, year_added, rating, duration, primary_country) |>
      datatable(options = list(pageLength = 15, scrollX = TRUE),
        filter = "top")
  })

  # ── OBSERVE: Reset button ─────────────────────────────
  observeEvent(input$reset, {
    updateSelectInput(session, "type", selected = "All")
    updateSliderInput(session, "years", value = c(2015, 2021))
    updateSelectInput(session, "country", selected = "All")
  })
}

# To run: shinyApp(app1_ui, app1_server)

# ══════════════════════════════════════════════════════════
# APP 2: e-Car LOAN EXPLORER
# Demonstrates: radioButtons, checkboxGroupInput, reactive, plotly
# ══════════════════════════════════════════════════════════

ec <- read_csv("ecar.csv") |>
  mutate(Year = year(mdy(`Approve Date`)),
    Spread = Rate - `Cost of Funds`) |>
  filter(Year >= 2002, Year <= 2012)

app2_ui <- fluidPage(
  titlePanel("e-Car Loan Explorer"),
  sidebarLayout(
    sidebarPanel(width = 3,
      # Credit tier selection
      checkboxGroupInput("tiers", "Credit Tiers:",
        choices = sort(unique(ec$Tier)),
        selected = sort(unique(ec$Tier))),

      # Year range
      sliderInput("ec_years", "Year Range:",
        min = 2002, max = 2012,
        value = c(2002, 2012), step = 1, sep = ""),

      # Metric selection
      radioButtons("metric", "Metric:",
        choices = c("Rate", "Spread", "Cost of Funds"),
        selected = "Rate"),

      textOutput("n_loans"),
      actionButton("ec_reset", "Reset")
    ),
    mainPanel(width = 9,
      tabsetPanel(
        tabPanel("Trend", plotlyOutput("ec_trend", height = "400px")),
        tabPanel("Distribution", plotOutput("ec_dist", height = "400px")),
        tabPanel("Data", DTOutput("ec_table"))
      )
    )
  )
)

app2_server <- function(input, output, session) {
  filtered_ec <- reactive({
    ec |>
      filter(Tier %in% input$tiers,
        Year >= input$ec_years[1],
        Year <= input$ec_years[2])
  })

  output$n_loans <- renderText({
    paste0("Showing ", format(nrow(filtered_ec()), big.mark = ","), " loans")
  })

  output$ec_trend <- renderPlotly({
    metric_col <- switch(input$metric,
      "Rate" = "Rate", "Spread" = "Spread", "Cost of Funds" = "Cost of Funds")
    yearly_ec <- filtered_ec() |>
      group_by(Year, Tier) |>
      summarise(value = mean(.data[[metric_col]], na.rm = TRUE), .groups = "drop")

    p <- ggplot(yearly_ec, aes(x = Year, y = value, color = factor(Tier),
        text = paste0("Tier ", Tier, "\nYear: ", Year,
          "\n", input$metric, ": ", round(value, 2)))) +
      geom_line(linewidth = 1) + geom_point(size = 2) +
      theme_minimal() +
      labs(y = input$metric, color = "Tier")
    ggplotly(p, tooltip = "text") |> layout(hovermode = "x unified")
  })

  output$ec_dist <- renderPlot({
    metric_col <- switch(input$metric,
      "Rate" = "Rate", "Spread" = "Spread", "Cost of Funds" = "Cost of Funds")
    ggplot(filtered_ec(), aes(x = .data[[metric_col]], fill = factor(Tier))) +
      geom_histogram(bins = 40, alpha = 0.6, position = "identity") +
      theme_minimal(base_size = 14) +
      labs(title = paste(input$metric, "Distribution by Tier"),
           x = input$metric, fill = "Tier")
  })

  output$ec_table <- renderDT({
    filtered_ec() |>
      select(Year, Tier, Rate, `Cost of Funds`, Spread, Outcome) |>
      datatable(options = list(pageLength = 20, scrollX = TRUE))
  })

  observeEvent(input$ec_reset, {
    updateCheckboxGroupInput(session, "tiers", selected = sort(unique(ec$Tier)))
    updateSliderInput(session, "ec_years", value = c(2002, 2012))
    updateRadioButtons(session, "metric", selected = "Rate")
  })
}

# To run: shinyApp(app2_ui, app2_server)

# ══════════════════════════════════════════════════════════
# INPUT WIDGET CHEATSHEET (printed)
# ══════════════════════════════════════════════════════════
cat("\n── Shiny Input Widget Cheatsheet ──\n")
widgets <- tibble(
  widget = c("selectInput", "sliderInput", "checkboxGroupInput",
    "radioButtons", "textInput", "actionButton",
    "dateRangeInput", "numericInput", "fileInput"),
  purpose = c("Single dropdown", "Numeric range", "Multiple checkboxes",
    "Single radio", "Free text", "Trigger event",
    "Date range", "Single number", "File upload"),
  shneiderman = c("Filter", "Filter", "Filter",
    "Reconfigure", "Filter", "History (reset)",
    "Filter", "Filter", "Extract (reverse)"))
print(widgets, n = 9)

cat("\n── Shiny Output/Render Pairs ──\n")
renders <- tibble(
  ui_function = c("plotOutput()", "plotlyOutput()", "DTOutput()",
    "textOutput()", "tableOutput()", "downloadButton()"),
  server_function = c("renderPlot()", "renderPlotly()", "renderDT()",
    "renderText()", "renderTable()", "downloadHandler()"),
  content = c("Static ggplot", "Interactive plotly", "Searchable DT table",
    "Text string", "Simple HTML table", "Downloadable file"))
print(renders, n = 6)

cat("\n── All W08-M04 R outputs saved ──\n")
cat("App 1: Netflix Explorer (run with shinyApp(app1_ui, app1_server))\n")
cat("App 2: e-Car Explorer (run with shinyApp(app2_ui, app2_server))\n")

File: latest/week3/workshop8_m4/W08_M04_demo.py
"""W08-M04: Shiny Fundamentals — Python companion
Shows the Dash equivalents for each Shiny pattern (preview for M05).
"""
import os; os.makedirs("output", exist_ok=True)

# ══════════════════════════════════════════════════════════
# SHINY ↔ DASH EQUIVALENCE TABLE
# (Full Dash tutorial in M05; this module is R-focused)
# ══════════════════════════════════════════════════════════

print("=== Shiny (R) ↔ Dash (Python) Equivalence ===\n")
equivalences = [
    ("ARCHITECTURE", "", ""),
    ("  fluidPage()", "app.layout = html.Div([...])", "Page container"),
    ("  sidebarLayout()", "dbc.Row([dbc.Col(...)])", "Layout grid"),
    ("  server <- function(input,output)", "@app.callback(...)", "Server logic"),
    ("", "", ""),
    ("INPUTS", "", ""),
    ("  selectInput('id', ...)", "dcc.Dropdown(id='id', ...)", "Dropdown"),
    ("  sliderInput('id', ...)", "dcc.RangeSlider(id='id', ...)", "Slider"),
    ("  checkboxGroupInput()", "dcc.Checklist(id='id', ...)", "Checkboxes"),
    ("  radioButtons()", "dcc.RadioItems(id='id', ...)", "Radio"),
    ("  actionButton()", "html.Button('Reset', id='id')", "Button"),
    ("", "", ""),
    ("OUTPUTS", "", ""),
    ("  plotOutput() + renderPlot()", "dcc.Graph(id='chart')", "Chart"),
    ("  DTOutput() + renderDT()", "dash_table.DataTable()", "Table"),
    ("  textOutput() + renderText()", "html.P(id='text')", "Text"),
    ("", "", ""),
    ("REACTIVITY", "", ""),
    ("  reactive({...})", "@app.callback(Output, Input)", "Auto-update"),
    ("  observeEvent(input$btn)", "@app.callback(..., Input('btn','n_clicks'))", "Event"),
    ("  eventReactive(input$go)", "Same callback, check n_clicks", "Lazy compute"),
]
for shiny, dash, note in equivalences:
    if not shiny: print(); continue
    print(f"  {shiny:<35} {dash:<40} {note}")

# ══════════════════════════════════════════════════════════
# MINIMAL DASH APP (preview for M05)
# ══════════════════════════════════════════════════════════
dash_template = '''
# ── MINIMAL DASH APP (W08-M05 preview) ──
# pip install dash pandas plotly

from dash import Dash, html, dcc, Input, Output
import pandas as pd
import plotly.express as px

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year

app = Dash(__name__)

app.layout = html.Div([
    html.H1("Netflix Content Explorer"),

    # Inputs
    html.Div([
        html.Label("Content Type:"),
        dcc.Dropdown(
            id="type-dropdown",
            options=[{"label": t, "value": t}
                     for t in ["All", "Movie", "TV Show"]],
            value="All"),

        html.Label("Year Range:"),
        dcc.RangeSlider(
            id="year-slider",
            min=2015, max=2021, step=1,
            value=[2015, 2021],
            marks={y: str(y) for y in range(2015, 2022)}),
    ], style={"width": "25%", "display": "inline-block",
              "verticalAlign": "top", "padding": "20px"}),

    # Output
    html.Div([
        dcc.Graph(id="trend-chart"),
        html.P(id="summary-text"),
    ], style={"width": "70%", "display": "inline-block"}),
])

@app.callback(
    [Output("trend-chart", "figure"),
     Output("summary-text", "children")],
    [Input("type-dropdown", "value"),
     Input("year-slider", "value")])
def update_chart(content_type, year_range):
    df = nf.query(f"{year_range[0]} <= year_added <= {year_range[1]}")
    if content_type != "All":
        df = df.query(f"type == '{content_type}'")

    yearly = df.groupby(["year_added", "type"]).size().reset_index(name="n")
    fig = px.line(yearly, x="year_added", y="n", color="type",
        color_discrete_map={"Movie": "#1565C0", "TV Show": "#E53935"},
        markers=True, title="Netflix Additions by Type")
    fig.update_layout(template="plotly_white", hovermode="x unified")

    summary = f"Showing {len(df):,} titles ({year_range[0]}–{year_range[1]})"
    return fig, summary

if __name__ == "__main__":
    app.run(debug=True)
'''

with open("output/dash_preview_app.py", "w") as f:
    f.write(dash_template)
print("\nDash preview app saved: output/dash_preview_app.py")
print("(Full Dash tutorial in W08-M05)")

# ══════════════════════════════════════════════════════════
# REACTIVITY COMPARISON
# ══════════════════════════════════════════════════════════
print("\n=== Reactivity: Shiny vs Dash ===")
print("  Shiny: reactive() = lazy, cached, auto-invalidates")
print("  Dash:  @callback  = triggers on any Input change")
print("")
print("  Shiny: observe()  = eager, side effects")
print("  Dash:  No direct equivalent (callbacks always return)")
print("")
print("  Shiny: eventReactive(input$btn) = trigger on click only")
print("  Dash:  Check n_clicks in callback; prevent_initial_call=True")
print("")
print("  Key difference: Shiny uses a REACTIVE GRAPH (implicit);")
print("  Dash uses EXPLICIT callbacks (you declare Input → Output).")
print("  Both achieve the same result; Shiny is more 'magical',")
print("  Dash is more explicit and debuggable.")

print("\nAll W08-M04 Python outputs saved")
